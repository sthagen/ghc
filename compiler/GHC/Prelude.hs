{-# LANGUAGE CPP #-}
{-# OPTIONS_HADDOCK not-home #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE ViewPatterns #-}

-- | Custom GHC "Prelude"
--
-- This module serves as a replacement for the "Prelude" module
-- and abstracts over differences between the bootstrapping
-- GHC version, and may also provide a common default vocabulary.

-- Every module in GHC
--   * Is compiled with -XNoImplicitPrelude
--   * Explicitly imports GHC.Prelude

module GHC.Prelude
  (module X
  ,module Bits
  ,shiftL, shiftR
#if __GLASGOW_HASKELL__ < 903
  ,SomeExceptionWithLocation
  ,pattern SomeExceptionWithLocation
#else
  ,SomeExceptionWithLocation(..)
#endif
  ) where


-- We export the 'Semigroup' class but w/o the (<>) operator to avoid
-- clashing with the (Outputable.<>) operator which is heavily used
-- through GHC's code-base.

{-
Note [Why do we import Prelude here?]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The files ghc-boot-th.cabal, ghc-boot.cabal, ghci.cabal and
ghc-heap.cabal contain the directive default-extensions:
NoImplicitPrelude. There are two motivations for this:
  - Consistency with the compiler directory, which enables
    NoImplicitPrelude;
  - Allows loading the above dependent packages with ghc-in-ghci,
    giving a smoother development experience when adding new
    extensions.
-}

import Prelude as X hiding ((<>))
import Data.Foldable as X (foldl')
#if __GLASGOW_HASKELL__ < 903
import Control.Exception ( SomeException(..) )
#else
import Control.Exception ( SomeExceptionWithLocation(..) )
#endif

#if MIN_VERSION_base(4,16,0)
import GHC.Bits as Bits hiding (shiftL, shiftR)
# if defined(DEBUG)
import qualified GHC.Bits as Bits (shiftL, shiftR)
# endif

#else
--base <4.15
import Data.Bits as Bits hiding (shiftL, shiftR)
# if defined(DEBUG)
import qualified Data.Bits as Bits (shiftL, shiftR)
# endif
#endif

{- Note [Default to unsafe shifts inside GHC]
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The safe shifts can introduce branches which come
at the cost of performance. We still want the additional
debugability for debug builds. So we define it as one or the
other depending on the DEBUG setting.

Why do we then continue on to re-export the rest of Data.Bits?
If we would not what is likely to happen is:
* Someone imports Data.Bits, uses xor. Things are fine.
* They add a shift and get an ambigious definition error.
* The are puzzled for a bit.
* They either:
  + Remove the import of Data.Bits and get an error because xor is not in scope.
  + Add the hiding clause to the Data.Bits import for the shifts.

Either is quite annoying. Simply re-exporting all of Data.Bits avoids this
making for a smoother developer experience. At the cost of having a few more
names in scope at all time. But that seems like a fair tradeoff.

See also #19618
-}

-- We always want the Data.Bits method to show up for rules etc.
{-# INLINE shiftL #-}
{-# INLINE shiftR #-}
shiftL, shiftR :: Bits.Bits a => a -> Int -> a
#if defined(DEBUG)
shiftL = Bits.shiftL
shiftR = Bits.shiftR
#else
shiftL = Bits.unsafeShiftL
shiftR = Bits.unsafeShiftR
#endif

#if __GLASGOW_HASKELL__ < 903
type SomeExceptionWithLocation = SomeException

{-# COMPLETE SomeExceptionWithLocation #-}
pattern SomeExceptionWithLocation :: forall. SomeException -> () -> SomeException
pattern SomeExceptionWithLocation e unit <- (\x -> ((), x) -> (unit, e))
  where
    SomeExceptionWithLocation (SomeException e) _ = SomeException e
#endif
