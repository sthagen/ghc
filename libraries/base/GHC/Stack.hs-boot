{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Stack (CallStack, HasCallStack, callStack) where

import GHC.Stack.Types

callStack :: HasCallStack => CallStack
