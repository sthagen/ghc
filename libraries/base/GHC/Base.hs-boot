{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Base (Maybe, Semigroup, Monoid, String) where

import GHC.Maybe (Maybe)
import GHC.Types (Char)

class Semigroup a
class Monoid a

type String = [Char]
