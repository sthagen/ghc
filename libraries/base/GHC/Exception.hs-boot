{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE ExistentialQuantification#-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeInType #-}

{-
This SOURCE-imported hs-boot module cuts a big dependency loop:

         GHC.Exception
imports  Data.Maybe
imports  GHC.Base
imports  GHC.Err
imports  {-# SOURCE #-} GHC.Exception

More dramatically

         GHC.Exception
imports  Data.Typeable
imports  Data.Typeable.Internal
imports  GHC.Arr (fingerprint representation etc)
imports  GHC.Real
imports  {-# SOURCE #-} GHC.Exception

However, GHC.Exceptions loop-breaking exports are all nice,
well-behaved, non-bottom values.  The clients use 'raise#'
to get a visibly-bottom value.
-}

module GHC.Exception
  ( module GHC.Exception.Type
  , errorCallException
  , errorCallWithCallStackException
  , throw
  ) where

import {-# SOURCE #-} GHC.Exception.Type
import GHC.Types ( Char, RuntimeRep, TYPE )
import GHC.Stack.Types ( CallStack, HasCallStack )

errorCallException :: [Char] -> SomeExceptionWithLocation
errorCallWithCallStackException :: [Char] -> CallStack -> SomeExceptionWithLocation

throw :: HasCallStack => forall (r :: RuntimeRep). forall (a :: TYPE r). forall e. Exception e => e -> a
