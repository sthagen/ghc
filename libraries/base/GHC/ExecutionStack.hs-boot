{-# LANGUAGE NoImplicitPrelude #-}

module GHC.ExecutionStack
  ( Location,
    getStackTrace,
  )
where

import GHC.Base
import {-# SOURCE #-} GHC.ExecutionStack.Internal (Location)

getStackTrace :: IO (Maybe [Location])
