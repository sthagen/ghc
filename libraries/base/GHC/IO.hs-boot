{-# LANGUAGE Unsafe #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.IO(
    IO(..),
    mplusIO,
    throwIOUserError
) where

import GHC.Types
import {-# SOURCE #-} GHC.Base (String)

mplusIO :: IO a -> IO a -> IO a
throwIOUserError :: String -> IO a
