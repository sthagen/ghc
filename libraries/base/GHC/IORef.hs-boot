{-# LANGUAGE NoImplicitPrelude #-}

module GHC.IORef where

import GHC.Base

data IORef a

newIORef :: a -> IO (IORef a)
readIORef :: IORef a -> IO a
writeIORef :: IORef a -> a -> IO ()
