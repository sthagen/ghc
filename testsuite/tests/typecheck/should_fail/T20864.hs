{-# language MagicHash, TransformListComp, GADTs, TypeFamilies #-}
module T20864 where
import GHC.Exts

-- This passes
loom :: [Int] -> [Int]
loom xs = [z | I# x <- xs, let p = x +# 3#, z <- [1 .. I# p], then take 3]


type family F a where
  F a = Int#

data Foo a where
  Foo :: F a -> F a -> Foo a
-- These fail
glum2 :: [Foo Int] -> [Int]
glum2 xs = [I# (x +# y) | Foo x y <- xs, then take 3]

glum :: [Int] -> [Int]
glum xs = [I# p | I# x <- xs, let p = x +# 3#, then take 3]
