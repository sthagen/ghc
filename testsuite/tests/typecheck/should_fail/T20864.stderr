
T20864.hs:17:33: error:
    • Expecting a lifted type, but got an unlifted type
    • In a stmt of a list comprehension: then take 3
      In the expression: [I# (x +# y) | Foo x y <- xs, then take 3]
      In an equation for ‘glum2’:
          glum2 xs = [I# (x +# y) | Foo x y <- xs, then take 3]

T20864.hs:20:22: error:
    • Expecting a lifted type, but got an unlifted type
    • In a stmt of a list comprehension: then take 3
      In the expression:
        [I# p | I# x <- xs, let p = x +# 3#, then take 3]
      In an equation for ‘glum’:
          glum xs = [I# p | I# x <- xs, let p = x +# 3#, then take 3]
